hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java

```diff
@@ -691,6 +691,9 @@ public void terminate(String reason, Exception cause, boolean clearMetrics,
       }
     }
 
+    if (this.replicationEndpoint != null) {
```

Comment:
ReplicationSourceShipper and entryReader are thread. replicationEndpoint is not a thread and is used by ReplicationSourceShipper thread. I am +1 for this: stop firstly, then interrupt thread.

Reply:
I think the intenion here to interrupt thread first is that, we want to make sure that, we could still do replication when shutting down the worker.
In general, I'm fine with this approach as in distributed system, we should have the ability to deal with this type error. But it is still a bit strange that, the above stop worker and setReaderRunning to false can not stop the work?

Reply:
> But it is still a bit strange that, the above stop worker and setReaderRunning to false can not stop the work?

Yes, stop worker and setReaderRunning to false just set flag non-running, if happens exception in while loop of method replicationEndpoint.replicate, will go into retry and cannot sense the flag changing. Leaving the retry can be achieved by calling replicationEndpoint.stop().